### PR TITLE
Skip graceful termination if Envoy ended prematurely

### PR DESCRIPTION
**Pull request by leosarra**
Currently, there is ambiguity in the behavior when both the pilot-agent and Envoy receive a SIGTERM signal "at around the same time". If the pilot-agent detects the SIGTERM before detecting the termination of the Envoy process, it initiates a graceful termination logic, waiting for X seconds. On the other, if it identifies the termination of the Envoy process before initiating the graceful logic, it terminates immediately.

This pull request change the behavior by skipping the remaining graceful waiting time in case Envoy terminates prematurely.
It's important to note that if the exitOnZeroActiveConnections flag is in use, this PR will not alter the behavior (but it will alter the logging messages). When exitOnZeroActiveConnections=true, the graceful termination logic already aborts prematurely as part of the error handling process.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
How does this happen? IIUC in k8s envoy will not get a sigterm on pod shutdown. Is it a manual signal? off-k8s?

**Comment by leosarra** [comment]
Yes, K8s is not expected to send SIGTERM to both processes, even during a node reboot only the main process shall receive the signal.
If the envoy process terminates early it usually means that the application is involved due to:
- the application container (with privileges) tinkering with other processes and signaling the envoy process
- Early termination via the /quitquitquit endpoint (perhaps triggered as part of the application termination logic)
Ofc, it could also happen that Envoy crashes as part of lifecycle but that would be quite abnormal occurrence.

Looking back I found this old ticket #36686 that mentioned the behavior with /quitquitquit.
At the time of the ticket the handling of exitOnZeroActiveConnections=true was also problematic (pilot-agent waiting indefinitely), that was later addressed via some improvements in the error handling.

File: pkg/envoy/agent.go [hunk 1]
```diff
@@ -175,31 +169,51 @@ func (a *Agent) terminate() {
...
-					" or the stats are not enabled. Skipping active connections check...")
-				a.abortCh <- errAbort
-				return
-			}
-			if ac == 0 {
-				log.Info("There are no more active connections. terminating proxy...")
-				a.abortCh <- errAbort
+			select {
+			case status := <-a.statusCh:
+				log.Warnf("Envoy exited with error %v", status.err)
```
Comment (hzxuzhonghu) [inline pkg/envoy/agent.go:176]: It is not always error

**Comment by leosarra** [comment]
> But doesn't agent have already exited immediately when envoy terminated
> 
> ```
> func (a *Agent) Run(ctx context.Context) {
> 	log.Info("Starting proxy agent")
> 	go a.runWait(a.abortCh)
> 
> 	select {
> 	case status := <-a.statusCh:
> 		if status.err != nil {
> 			if status.err.Error() == errOutOfMemory {
> 				log.Warnf("Envoy may have been out of memory killed. Check memory usage and limits.")
> 			}
> 			log.Errorf("Envoy exited with error: %v", status.err)
> 		} else {
> 			log.Infof("Envoy exited normally")
> 		}
> ```

Yes pilot-agent will exit early but only if it did not start the graceful termination process already.

**Comment by leosarra** [comment]
I'm not sure if a release-note entry is needed for this change.
I would prefer to add one in order to link the changes to #36686 but I cannot find a "fitting area" where this commit belongs to :P
Do you have any proposals?

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**